### PR TITLE
Combine Error with Reason when Connection State Changed

### PR DESCRIPTION
**Pull request by anthony-murphy**
Looking at the logs there are a number of cases where we see ConnectionStateChange events which have values that imply an error happened, but the error itself is not logged. This change couples error and reason via typing to prevent losing one or the other duration connection state changes.

### CODE REVIEW

File: packages/loader/container-loader/src/connectionManager.ts [hunk 1]
```diff
@@ -956,9 +964,13 @@ export class ConnectionManager implements IConnectionManager {
...
 		this.triggerConnect(
-			error !== undefined
-				? "Reconnecting due to Error"
-				: `Reconnecting due to: ${disconnectMessage}`,
+			{
+				description:
+					reason.error !== undefined
+						? "Reconnecting due to Error"
+						: `Reconnecting due to: ${reason.description}`,
+				error: reason.error,
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/connectionManager.ts:972]: this is a place we were for sure losing an error, there are a few other suspicious spots, but frankly, the code is hard to follow, so many layers with the same methods names, and lots of passing between them. Hence the reason i went with this approach, as it was easier to ensure the right pieces get passed along.

**Comment by anthony-murphy** [comment]
Note: this is on next, as to do it in main would be a breaking change, as a number of impacted types were deprecated and removed this release cycle.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 2]
```diff
@@ -347,6 +350,7 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
 		return {
 			sequenceNumber: this.lastSequenceNumber,
 			opsSize: this.opsSize > 0 ? this.opsSize : undefined,
+			deltaManagerState: this._disposed ? "disposed" : this._closed ? "closed" : "open",
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/deltaManager.ts:353]: this will help differentiate closes due to dispose/close from disconnects for other reasons

File: packages/loader/container-loader/src/container.ts [hunk 3]
```diff
@@ -1338,23 +1342,23 @@ export class Container
...
-		this.setAutoReconnectInternal(mode);
+		this.setAutoReconnectInternal(mode, args.reason);
 	}
 
 	public disconnect() {
 		if (this.closed) {
 			throw new UsageError(`The Container is closed and cannot be disconnected`);
 		} else {
-			this.disconnectInternal();
+			this.disconnectInternal({ text: "DocumentDisconnect" });
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:1352]: previously these would show up as reason setAutoReconnect, which was isn't very useful, now all the code paths set a distinct reason

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 4]
```diff
@@ -150,15 +148,15 @@ class ConnectionStateHandlerPassThrough
 	public initProtocol(protocol: IProtocolHandler) {
 		return this.pimpl.initProtocol(protocol);
 	}
-	public receivedDisconnectEvent(reason: string, error?: IAnyDriverError) {
-		return this.pimpl.receivedDisconnectEvent(reason, error);
+	public receivedDisconnectEvent(reason: IConnectionStateChangeReason) {
```
Comment (wes-carlson) [inline packages/loader/container-loader/src/connectionStateHandler.ts:151]: Should this also be `IConnectionStateChangeReason<IAnyDriverError>`?

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 5]
```diff
@@ -176,10 +174,9 @@ class ConnectionStateHandlerPassThrough
 	public connectionStateChanged(
 		value: ConnectionState,
 		oldState: ConnectionState,
-		reason?: string | undefined,
-		error?: IAnyDriverError,
+		reason?: IConnectionStateChangeReason,
```
Comment (wes-carlson) [inline packages/loader/container-loader/src/connectionStateHandler.ts:177]: same question

File: packages/loader/container-loader/src/deltaManager.ts (hunk 2 above)
Comment (wes-carlson) [reply packages/loader/container-loader/src/deltaManager.ts:353]: Don't the reason strings already do that? Won't this always be the same as the container state?

**Review by wes-carlson — APPROVED** [review]

File: packages/loader/container-loader/src/deltaManager.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/deltaManager.ts:353]: no. during disposal it get weird. The first thing that happens is dispose on the dm, which disconnects, but only after dm disposed finishes does the container dispose kick in. in the logs it looks like:
<HTML><HEAD><TITLE>From Clipboard</TITLE></HEAD><BODY>

Data_reason | Data_containerLifecycleState | Data_containerConnectionState
-- | -- | -- | -- | --
Closing DeltaManager |  loaded | Disconnected


</BODY></HTML>

File: packages/loader/container-loader/src/connectionManager.ts [hunk 6]
```diff
@@ -481,14 +485,17 @@ export class ConnectionManager implements IConnectionManager {
 		}
 	}
 
-	public connect(reason: string, connectionMode?: ConnectionMode) {
-		this.connectCore(reason, connectionMode).catch((error) => {
-			const normalizedError = normalizeError(error, { props: fatalConnectErrorProp });
+	public connect(reason: IConnectionStateChangeReason, connectionMode?: ConnectionMode) {
```
[the remark this answers is not in the document]
Comment (anthony-murphy) [reply packages/loader/container-loader/src/connectionManager.ts:488]: its cheaper to just have a reason the flows all the way through.

File: packages/loader/container-loader/src/connectionStateHandler.ts (hunk 4 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/connectionStateHandler.ts:151]:
```suggestion
	public receivedDisconnectEvent(reason: IConnectionStateChangeReason<IAnyDriverError>) {
```

File: packages/loader/container-loader/src/connectionStateHandler.ts (hunk 5 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/connectionStateHandler.ts:177]:
```suggestion
		reason?: IConnectionStateChangeReason<IAnyDriverError>,
```

File: packages/loader/container-loader/src/connectionManager.ts [hunk 7]
```diff
@@ -684,7 +691,7 @@ export class ConnectionManager implements IConnectionManager {
 	 * @param error - Error causing the disconnect if any.
 	 * @returns A boolean that indicates if there was an existing connection (or pending connection) to disconnect
 	 */
-	private disconnectFromDeltaStream(reason: string, error?: IAnyDriverError): boolean {
+	private disconnectFromDeltaStream(reason: IConnectionStateChangeReason): boolean {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/connectionManager.ts:694]:
```suggestion
	private disconnectFromDeltaStream(reason: IConnectionStateChangeReason<IAnyDriverError>): boolean {
```

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 8]
```diff
@@ -33,8 +32,7 @@ export interface IConnectionStateHandlerInputs {
 	connectionStateChanged: (
 		value: ConnectionState,
 		oldState: ConnectionState,
-		reason?: string | undefined,
-		error?: IAnyDriverError,
+		reason?: IConnectionStateChangeReason,
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/connectionStateHandler.ts:35]:
```suggestion
		reason?: IConnectionStateChangeReason<IAnyDriverError>,
```

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 9]
```diff
@@ -61,14 +59,14 @@ export interface IConnectionStateHandler {
 	dispose(): void;
 	initProtocol(protocol: IProtocolHandler): void;
 	receivedConnectEvent(details: IConnectionDetailsInternal): void;
-	receivedDisconnectEvent(reason: string, error?: IAnyDriverError): void;
-	establishingConnection(reason: string): void;
+	receivedDisconnectEvent(reason: IConnectionStateChangeReason): void;
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/connectionStateHandler.ts:62]:
```suggestion
	receivedDisconnectEvent(reason: IConnectionStateChangeReason<IAnyDriverError>): void;
```

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 10]
```diff
@@ -223,8 +220,7 @@ class ConnectionStateCatchup extends ConnectionStateHandlerPassThrough {
 	public connectionStateChanged(
 		value: ConnectionState,
 		oldState: ConnectionState,
-		reason?: string | undefined,
-		error?: IAnyDriverError,
+		reason?: IConnectionStateChangeReason,
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/connectionStateHandler.ts:223]:
```suggestion
		reason?: IConnectionStateChangeReason<IAnyDriverError>,
```

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 11]
```diff
@@ -583,14 +576,12 @@ class ConnectionStateHandler implements IConnectionStateHandler {
...
-		reason: string,
-		error?: IAnyDriverError,
+		reason: IConnectionStateChangeReason,
 	): void;
 	private setConnectionState(value: ConnectionState.Connected): void;
 	private setConnectionState(
 		value: ConnectionState.Disconnected | ConnectionState.Connected,
-		reason?: string,
-		error?: IAnyDriverError,
+		reason?: IConnectionStateChangeReason,
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/connectionStateHandler.ts:584]:
```suggestion
		reason?: IConnectionStateChangeReason<IAnyDriverError>,
```

File: packages/loader/container-loader/src/container.ts [hunk 12]
```diff
@@ -2070,8 +2074,7 @@ export class Container
 	private logConnectionStateChangeTelemetry(
 		value: ConnectionState,
 		oldState: ConnectionState,
-		reason?: string,
-		error?: IAnyDriverError,
+		reason?: IConnectionStateChangeReason,
```
[the remark this answers is not in the document]
Comment (anthony-murphy) [reply packages/loader/container-loader/src/container.ts:2077]:
```suggestion
		reason?: IConnectionStateChangeReason<IAnyDriverError>,
```

File: packages/loader/container-loader/src/contracts.ts [hunk 13]
```diff
@@ -150,7 +156,7 @@ export interface IConnectionManagerFactoryArgs {
 	/**
 	 * Called whenever connection to relay service is lost.
 	 */
-	readonly disconnectHandler: (reason: string, error?: IAnyDriverError) => void;
+	readonly disconnectHandler: (reason: IConnectionStateChangeReason) => void;
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/contracts.ts:159]:
```suggestion
	readonly disconnectHandler: (reason: IConnectionStateChangeReason<IAnyDriverError>) => void;
```

File: packages/loader/container-loader/src/deltaManager.ts [hunk 14]
```diff
@@ -400,15 +404,17 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
 			reconnectionDelayHandler: (delayMs: number, error: unknown) =>
 				this.emitDelayInfo(this.deltaStreamDelayId, delayMs, error),
 			closeHandler: (error: any) => this.close(error),
-			disconnectHandler: (reason: string, error?: IAnyDriverError) =>
-				this.disconnectHandler(reason, error),
+			disconnectHandler: (reason: IConnectionStateChangeReason) =>
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/deltaManager.ts:407]:
```suggestion
			disconnectHandler: (reason: IConnectionStateChangeReason<IAnyDriverError>) =>
```

File: packages/loader/container-loader/src/connectionManager.ts [hunk 15]
```diff
@@ -406,7 +410,7 @@ export class ConnectionManager implements IConnectionManager {
 	 * Enables or disables automatic reconnecting.
 	 * Will throw an error if reconnectMode set to Never.
 	 */
-	public setAutoReconnect(mode: ReconnectMode): void {
+	public setAutoReconnect(mode: ReconnectMode, reason: IConnectionStateChangeReason): void {
```
[the remark this answers is not in the document]
Comment (anthony-murphy) [reply packages/loader/container-loader/src/connectionManager.ts:413]: it should always be specified, so we know why the connection state changes, and i don't want to preclude an error being part of that reason

**Comment by anthony-murphy** [comment]
resolves AB#5197

**Comment by kian-thompson** [comment]
@anthony-murphy Doesn't this break the contract on `IDeltaManagerEvents`? A number of events had their shape changed.

**Comment by anthony-murphy** [comment]
> @anthony-murphy Doesn't this break the contract on `IDeltaManagerEvents`? A number of events had their shape changed.

@kian-thompson  can you provide an example? nothing should have change besides internals

**Comment by anthony-murphy** [comment]
> > @anthony-murphy Doesn't this break the contract on `IDeltaManagerEvents`? A number of events had their shape changed.
> 
> @kian-thompson can you provide an example? nothing should have change besides internals. This was also released as part of client_v2.0.0-internal.6.1.0, so not sure how it could be causing an issue.

**Comment by kian-thompson** [comment]
> @kian-thompson can you provide an example?

`DeltaManager.emit("disconnect", ...)` will emit an object of shape `IConnectionStateChangeReason`.
https://github.com/microsoft/FluidFramework/blob/main/packages/loader/container-loader/src/deltaManager.ts#L800

If I use `IContainer.deltaManager` and try listening to the `"disconnect"` event, I get an `IConnectionStateChangeReason` object when I expect `(reason: string, error?: IAnyDriverError)` based on the definition in `IDeltaManagerEvents`.
https://github.com/microsoft/FluidFramework/blob/main/packages/common/container-definitions/src/deltas.ts#L120

**Comment by anthony-murphy** [comment]
> > @kian-thompson can you provide an example?
> 
> `DeltaManager.emit("disconnect", ...)` will emit an object of shape `IConnectionStateChangeReason`. https://github.com/microsoft/FluidFramework/blob/main/packages/loader/container-loader/src/deltaManager.ts#L800
> 
> If I use `IContainer.deltaManager` and try listening to the `"disconnect"` event, I get an `IConnectionStateChangeReason` object when I expect `(reason: string, error?: IAnyDriverError)` based on the definition in `IDeltaManagerEvents`. https://github.com/microsoft/FluidFramework/blob/main/packages/common/container-definitions/src/deltas.ts#L120

good catch. seems like a bug. fix should be trivial, just need to pull text and error off the object and pass those to the emit method instead of the object

**Comment by kian-thompson** [comment]
> good catch. seems like a bug. fix should be trivial, just need to pull text and error off the object and pass those to the emit method instead of the object

[AB#7123](https://dev.azure.com/fluidframework/internal/_workitems/edit/7123)
